### PR TITLE
Demangling: Assert child count in mangleOpaqueType

### PR DESCRIPTION
**Pull request by kastiglione**
rdar://88080388

### CODE REVIEW

**Comment by kastiglione** [comment]
@swift-ci test

**Review by adrian-prantl — APPROVED** [review]
This change looks obviously correct. It would be good to create a unit test binary similar to LLDB's https://github.com/apple/llvm-project/blob/f22094d4f79ffc00c47abaa002327f89122f96ee/lldb/unittests/Symbol/TestTypeSystemSwiftTypeRef.cpp#L36 so we can add test coverage for this.

Does the DEMANGLER_ASSERT macro return on error? Only in an NDEBUG build?

**Comment by kastiglione** [comment]
I'll add a test.

> Does the `DEMANGLER_ASSERT` macro return on error? Only in an `NDEBUG` build?

correct on both.
 https://github.com/apple/swift/blob/eb6f9e0bbaf3255a9ea4161fcd0c08cdd15ba662/lib/Demangling/DemanglerAssert.h#L25-L32

**Comment by adrian-prantl** [comment]
> I'll add a test.
> 
> > Does the `DEMANGLER_ASSERT` macro return on error? Only in an `NDEBUG` build?
> 
> correct on both.

Hmm.. that will make writing a test for it a challenge though. Should we have a different macro that doesn't   have an assertion and returns an error instead? Or make the assertion behavior configurable somehow, so it can be disabled when running from a unit test only?

**Comment by kastiglione** [comment]
@adrian-prantl Given that a test will be in llvm-project, and that it's not immediately clear how to construct the test, do you think it makes sense to merge this ahead of the test? Or should it wait for a test?

**Comment by kastiglione** [comment]
Spoke with Adrian offline, will merge now and follow up with a test.

**Review by fredriss — APPROVED** [review]

File: lib/Demangling/Remangler.cpp [hunk 1]
```diff
@@ -3327,6 +3327,7 @@ ManglingError Remangler::mangleOpaqueType(Node *node, unsigned depth) {
   if (trySubstitution(node, entry))
     return ManglingError::Success;
 
+  DEMANGLER_ASSERT(node->getNumChildren() >= 2, node);
```
Comment (kastiglione) [inline lib/Demangling/Remangler.cpp:3330]: Fred pointed out this should be `3` not two (below `node->getChild(2)` is accessed).

**Pull request merged**